File: launcher4j/engine/file_watcher.py
"""File system watcher for auto-compilation."""

import os
import threading
import time
from typing import Callable, Optional

from watchdog.events import FileSystemEventHandler
from watchdog.observers import Observer


class JavaFileHandler(FileSystemEventHandler):
    """Watchdog handler that triggers on .java file changes."""

    def __init__(
        self,
        project_id: str,
        project_path: str,
        debounce_ms: int,
        on_change: Callable[[str], None],
    ):
        self.project_id = project_id
        self.project_path = project_path
        self.debounce_ms = debounce_ms
        self.on_change = on_change
        self._timer: Optional[threading.Timer] = None
        self._lock = threading.Lock()

    def on_modified(self, event):
        if event.is_directory:
            return
        if not event.src_path.endswith(".java"):
            return

        # Debounce: reset timer on each change
        with self._lock:
            if self._timer and self._timer.is_alive():
                self._timer.cancel()
            self._timer = threading.Timer(
                self.debounce_ms / 1000.0,
                self._fire,
            )
            self._timer.daemon = True
            self._timer.start()

    def _fire(self):
        self.on_change(self.project_id)


class FileWatcher:
    """Watches project src/ directories for .java file changes."""

    def __init__(self):
        self._observers: dict[str, tuple[Observer, JavaFileHandler]] = {}
        self._lock = threading.Lock()

    def watch(
        self,
        project_id: str,
        project_path: str,
        debounce_ms: int = 500,
        on_compile: Optional[Callable[[str], None]] = None,
    ) -> bool:
        src_dir = os.path.join(project_path, "src")
        if not os.path.isdir(src_dir):
            return False

        self.unwatch(project_id)

        handler = JavaFileHandler(project_id, project_path, debounce_ms,
                                   lambda pid: on_compile(pid) if on_compile else None)
        observer = Observer()
        observer.schedule(handler, src_dir, recursive=True)
        observer.daemon = True
        observer.start()

        with self._lock:
            self._observers[project_id] = (observer, handler)

        return True

    def unwatch(self, project_id: str):
        with self._lock:
            if project_id in self._observers:
                observer, handler = self._observers.pop(project_id)
                observer.stop()
                observer.join(timeout=3)

    def unwatch_all(self):
        with self._lock:
            for project_id in list(self._observers.keys()):
                self.unwatch(project_id)

    def update_debounce(self, project_id: str, debounce_ms: int):
        with self._lock:
            if project_id in self._observers:
                _, handler = self._observers[project_id]
                handler.debounce_ms = debounce_ms

File: internal/engine/filewatcher.go
package engine

import (
	"os"
	"path/filepath"
	"strings"
	"sync"
	"time"

	"github.com/fsnotify/fsnotify"
)

// FileWatcher watches project src/ directories for .java file changes
// and triggers debounced compile callbacks.
type FileWatcher struct {
	mu        sync.Mutex
	watchers  map[string]*projectWatcher
}

type projectWatcher struct {
	watcher    *fsnotify.Watcher
	projectID  string
	projectPath string
	debounceMs int
	onChange   func(string)
	timer      *time.Timer
	mu         sync.Mutex
	done       chan struct{}
}

// NewFileWatcher creates a new FileWatcher.
func NewFileWatcher() *FileWatcher {
	return &FileWatcher{
		watchers: make(map[string]*projectWatcher),
	}
}

// Watch starts watching a project's src/ directory for .java file changes.
// Returns true if watching was started successfully.
func (fw *FileWatcher) Watch(projectID, projectPath string, debounceMs int, onChange func(string)) bool {
	srcDir := filepath.Join(projectPath, "src")
	info, err := os.Stat(srcDir)
	if err != nil || !info.IsDir() {
		return false
	}

	fw.Unwatch(projectID)

	watcher, err := fsnotify.NewWatcher()
	if err != nil {
		return false
	}

	pw := &projectWatcher{
		watcher:     watcher,
		projectID:   projectID,
		projectPath: projectPath,
		debounceMs:  debounceMs,
		onChange:    onChange,
		done:        make(chan struct{}),
	}

	fw.mu.Lock()
	fw.watchers[projectID] = pw
	fw.mu.Unlock()

	// Add all subdirectories recursively
	filepath.Walk(srcDir, func(path string, fi os.FileInfo, err error) error {
		if err != nil {
			return nil
		}
		if fi.IsDir() {
			watcher.Add(path)
		}
		return nil
	})

	go pw.loop()

	return true
}

// Unwatch stops watching a project.
func (fw *FileWatcher) Unwatch(projectID string) {
	fw.mu.Lock()
	pw, ok := fw.watchers[projectID]
	if ok {
		delete(fw.watchers, projectID)
	}
	fw.mu.Unlock()

	if ok {
		close(pw.done)
		pw.watcher.Close()
	}
}

// UnwatchAll stops all watchers.
func (fw *FileWatcher) UnwatchAll() {
	fw.mu.Lock()
	ids := make([]string, 0, len(fw.watchers))
	for id := range fw.watchers {
		ids = append(ids, id)
	}
	fw.mu.Unlock()

	for _, id := range ids {
		fw.Unwatch(id)
	}
}

// UpdateDebounce updates the debounce interval for a project watcher.
func (fw *FileWatcher) UpdateDebounce(projectID string, debounceMs int) {
	fw.mu.Lock()
	defer fw.mu.Unlock()
	if pw, ok := fw.watchers[projectID]; ok {
		pw.mu.Lock()
		pw.debounceMs = debounceMs
		pw.mu.Unlock()
	}
}

func (pw *projectWatcher) loop() {
	for {
		select {
		case event, ok := <-pw.watcher.Events:
			if !ok {
				return
			}
			if event.Op&(fsnotify.Create|fsnotify.Write) == 0 {
				continue
			}
			if !strings.HasSuffix(event.Name, ".java") {
				continue
			}
			pw.debounceFire()

		case err, ok := <-pw.watcher.Errors:
			if !ok {
				return
			}
			_ = err // silently ignore watch errors

		case <-pw.done:
			return
		}
	}
}

func (pw *projectWatcher) debounceFire() {
	pw.mu.Lock()
	defer pw.mu.Unlock()

	if pw.timer != nil {
		pw.timer.Stop()
	}

	pw.timer = time.AfterFunc(time.Duration(pw.debounceMs)*time.Millisecond, func() {
		if pw.onChange != nil {
			pw.onChange(pw.projectID)
		}
	})
}
